File: src/main/cwd.ts
import { execFile } from 'node:child_process';
import { promises as fs } from 'node:fs';

// macOS has no /proc; ask lsof for the process's cwd. Returns null if lsof is
// missing or the pid is gone.
function lsofCwd(pid: number): Promise<string | null> {
  return new Promise((resolve) => {
    execFile('lsof', ['-a', '-p', String(pid), '-d', 'cwd', '-Fn'], (err, stdout) => {
      if (err) return resolve(null);
      // The line starting with 'n' carries the cwd path.
      const line = stdout.split('\n').find((l) => l.startsWith('n'));
      resolve(line && line.length > 1 ? line.slice(1) : null);
    });
  });
}

// Resolve the live cwd of a process — `/proc/<pid>/cwd` on Linux, lsof on macOS.
// Returns null when there's nothing to report (bad pid, gone, or unreadable).
export async function liveCwd(pid: number | undefined): Promise<string | null> {
  if (!pid || pid <= 0) return null;
  try {
    const link = await fs.readlink(`/proc/${pid}/cwd`);
    if (link) return link;
  } catch {
    // not linux, or unreadable — fall through to lsof
  }
  return lsofCwd(pid);
}

File: src-tauri/src/cwd.rs
//! 对偶 src/main/cwd.ts。macOS 走 lsof（无 /proc），Linux 走 readlink /proc。

use std::path::PathBuf;

/// 解析进程的实时 cwd。None = 读不到（pid 无效或进程已退）。
pub fn live_cwd(pid: u32) -> Option<PathBuf> {
    if pid == 0 {
        return None;
    }
    #[cfg(target_os = "linux")]
    {
        if let Ok(p) = std::fs::read_link(format!("/proc/{}/cwd", pid)) {
            return Some(p);
        }
    }
    lsof_cwd(pid)
}

// macOS 无 /proc；问 lsof 该进程的 cwd。lsof 缺失或 pid 已退 → None。
fn lsof_cwd(pid: u32) -> Option<PathBuf> {
    let output = std::process::Command::new("lsof")
        .args(["-a", "-p", &pid.to_string(), "-d", "cwd", "-Fn"])
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let stdout = String::from_utf8_lossy(&output.stdout);
    // 'n' 开头的行携带 cwd 路径
    stdout
        .lines()
        .find(|l| l.starts_with('n'))
        .and_then(|l| l.strip_prefix('n'))
        .map(PathBuf::from)
}
